### ISSUE TITLE
YOLO11-seg training takes up more GPU memory than YOLOv8-seg

### ISSUE THREAD

**Issue by panp4n** [ultralytics/ultralytics#16757]
### Search before asking

- [X] I have searched the Ultralytics YOLO [issues](https://github.com/ultralytics/ultralytics/issues) and [discussions](https://github.com/ultralytics/ultralytics/discussions) and found no similar questions.

### Question

training parameters:
model: YOLO11s-seg\YOLOv8s-seg
the same batchsize: 3
the same dataset
Other training parameters are consistent

YOLO11-seg training takes up more about 6GB GPU memory than YOLOv8-seg.

**Comment by UltralyticsAssistant** [comment]
👋 Hello @panp4n, thank you for your interest in Ultralytics 🚀! 

We recommend checking out our [Docs](https://docs.ultralytics.com) where you can find many [Python](https://docs.ultralytics.com/usage/python/) and [CLI](https://docs.ultralytics.com/usage/cli/) examples. They might have answers to common questions.

For your 🐛 Bug Report, please provide a [minimum reproducible example](https://docs.ultralytics.com/help/minimum_reproducible_example/) to help us debug the issue you're experiencing with YOLO11-seg.

If you have custom training ❓ Questions, please give more details, such as dataset image examples and training logs. Also, ensure you're following our [Tips for Best Training Results](https://docs.ultralytics.com/guides/model-training-tips/).

Consider joining our community! For real-time chat, visit [Discord](https://ultralytics.com/discord) 🎧. Prefer discussions? Head over to [Discourse](https://community.ultralytics.com), or check out our [Subreddit](https://reddit.com/r/ultralytics).

## Upgrade

Make sure you're using the latest version by upgrading the `ultralytics` package along with all [requirements](https://github.com/ultralytics/ultralytics/blob/main/pyproject.toml) in a [**Python>=3.8**](https://www.python.org/) environment with [**PyTorch>=1.8**](https://pytorch.org/get-started/locally/):

```bash
pip install -U ultralytics
```

## Environments

YOLO runs well in various environments with dependencies like [CUDA](https://developer.nvidia.com/cuda)/[CUDNN](https://developer.nvidia.com/cudnn):

- **Notebooks** with free GPU support: [Paperspace](https://console.paperspace.com/github/ultralytics/ultralytics), [Colab](https://colab.research.google.com/github/ultralytics/ultralytics/blob/main/examples/tutorial.ipynb), [Kaggle](https://www.kaggle.com/ultralytics/yolov8)
- **Google Cloud** Deep Learning VM. See [GCP Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/google_cloud_quickstart_tutorial/)
- **Amazon** Deep Learning AMI. See [AWS Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/aws_quickstart_tutorial/)
- **Docker Image**. See [Docker Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/docker_image_quickstart_tutorial/) [Docker Hub](https://hub.docker.com/r/ultralytics/ultralytics)

## Status

<a href="https://github.com/ultralytics/ultralytics/actions/workflows/ci.yaml?query=event%3Aschedule"><img src="https://github.com/ultralytics/ultralytics/actions/workflows/ci.yaml/badge.svg" alt="Ultralytics CI"></a>

If this badge is green, all [Ultralytics CI](https://github.com/ultralytics/ultralytics/actions/workflows/ci.yaml?query=event%3Aschedule) tests are passing, which confirms proper operation across different systems.

This is an automated response; an Ultralytics engineer will assist you soon.

**Comment by R-C-K-a** [comment]
https://github.com/ultralytics/ultralytics/blob/5dcaa0aa06ad580d434a7adfe145fab13354ab5d/ultralytics/nn/modules/block.py#L948C9-L948C77

**Comment by Y-T-G** [comment]
I can't reproduce this in the latest version. Please update your ultralytics.

**Comment by Y-T-G** [comment]
I can't reproduce this in the latest version. Please update your ultralytics.

**Comment by glenn-jocher** [comment]
Great to hear it's working now! If you have any more questions, feel free to ask.

**Comment by panp4n** [comment]
Another question:
There are extra parts at the edge of the mask in YOLO11-seg predict. 
![image](https://github.com/user-attachments/assets/fb423a07-99a4-408d-8199-0703dea58638)

In my own inference code, it is fine.
![image](https://github.com/user-attachments/assets/6b9a4aa7-1cda-4f56-9ebe-90e65b2c0668)

In YOLOv8-seg predict, it is also fine.
![image](https://github.com/user-attachments/assets/e1c8bf0c-14c0-4a80-aca9-0ca759c2a098)
I think YOLO11-seg predict script has some difference for mask processing.

**Comment by Y-T-G** [comment]
You can try passing `retina_masks=True`

**Comment by panp4n** [comment]
> You can try passing `retina_masks=True`

It works, Thank you so much!
![image](https://github.com/user-attachments/assets/b0fe543a-2417-4f24-a965-1dff1a637cb0)

I need optimize my inference code, How can I implement retina_masks  in my code.

**Comment by glenn-jocher** [comment]
@panp4n you're welcome! To implement `retina_masks` in your code, you can set the parameter when calling the model's predict function, like this: `model.predict(source, retina_masks=True)`.

**Comment by panp4n** [comment]
> @panp4n you're welcome! To implement `retina_masks` in your code, you can set the parameter when calling the model's predict function, like this: `model.predict(source, retina_masks=True)`.

I knew that, I want to know how to implement mask post-process like `retina_masks` in the onnxruntime or openvino inference.
As you can see, the edge of this mask is rough.
![image](https://github.com/user-attachments/assets/c6739c16-d2a2-4258-9466-8550586b5652)

**Comment by Y-T-G** [comment]
The code used by `retina_masks` is here

https://github.com/ultralytics/ultralytics/blob/8154a27e9b332dd2dac7a107003462f0e20f6179/ultralytics/utils/ops.py#L689

https://github.com/ultralytics/ultralytics/blob/8154a27e9b332dd2dac7a107003462f0e20f6179/ultralytics/models/yolo/segment/predict.py#L49-L50

**Comment by glenn-jocher** [comment]
You're welcome! If you have any more questions, feel free to ask.

**Comment by panp4n** [comment]
> The code used by `retina_masks` is here
> 
> https://github.com/ultralytics/ultralytics/blob/8154a27e9b332dd2dac7a107003462f0e20f6179/ultralytics/utils/ops.py#L689
> 
> https://github.com/ultralytics/ultralytics/blob/8154a27e9b332dd2dac7a107003462f0e20f6179/ultralytics/models/yolo/segment/predict.py#L49-L50

https://github.com/ultralytics/ultralytics/blob/8154a27e9b332dd2dac7a107003462f0e20f6179/ultralytics/utils/ops.py#L729
I can't reproduce this `align_corners=False` by using OpenCV C++ api .
The example also uses `cv2.resize()`,  My code is similar to this. There is still a gap in effectiveness.
https://github.com/ultralytics/ultralytics/blob/8154a27e9b332dd2dac7a107003462f0e20f6179/examples/YOLOv8-Segmentation-ONNXRuntime-Python/main.py#L256-L258

**Comment by panp4n** [comment]
The same model, why the mask of official predict script is better than deployment code?
Do you have any recommendations?

**Comment by panp4n** [comment]
C++ code is as follows:
`for (int i = 0; i < output_buffer.rows; i++) {
        cv::Mat classes_scores = output_buffer.row(i).colRange(4, output_buffer.cols-32);
        cv::Point class_id;
        double maxClassScore;
        cv::minMaxLoc(classes_scores, 0, &maxClassScore, 0, &class_id);

        if (maxClassScore > scoreThreshold_) {
            class_scores.push_back(maxClassScore);
            class_ids.push_back(class_id.x);
            float cx = output_buffer.at<float>(i, 0);
            float cy = output_buffer.at<float>(i, 1);
            float w = output_buffer.at<float>(i, 2);
            float h = output_buffer.at<float>(i, 3);
            int left = round((cx - 0.5 * w));
            int top = round((cy - 0.5 * h));
            int width = round(w);
            int height = round(h);
            cv::Mat mask_conf = output_buffer.row(i).colRange(output_buffer.cols - 32, output_buffer.cols);
            mask_confs.push_back(mask_conf);
            boxes.push_back(cv::Rect(left, top, width, height));
        }
    }
    //NMS
    std::vector<int> indices;
    cv::dnn::NMSBoxes(boxes, class_scores, scoreThreshold_, nms_threshold_, indices);
    cv::Mat rgb_mask = cv::Mat::zeros(image.size(), image.type());
    cv::Mat masked_img;
    cv::RNG rng;
    for (size_t i = 0; i < indices.size(); i++)
    {
        InferResult res;
        int index = indices[i];
        int class_id = class_ids[index];
        res.labelIdx = class_id;
        res.labelName = labelVect_[class_id];
        cv::Mat m = mask_confs[index] * proto;
        for (int col = 0; col < m.cols; col++) {
            m.at<float>(0, col) = sigmoid_function(m.at<float>(0, col));
        }
        cv::Mat m1 = m.reshape(1, sqrt(proto.cols)); // 1x25600 -> 160x160
        int x1 = (std::max)(0, boxes[index].x);
        int y1 = (std::max)(0, boxes[index].y);
        int x2 = (std::max)(0, boxes[index].br().x);
        int y2 = (std::max)(0, boxes[index].br().y);
        int mx1 = round(x1 * 0.25);
        int my1 = round(y1 * 0.25);
        int mx2 = round(x2 * 0.25);
        int my2 = round(y2 * 0.25);
        boxes[index].x = round(boxes[index].x * scale_factor_);
        boxes[index].y = round(boxes[index].y * scale_factor_);
        boxes[index].width = round(boxes[index].width * scale_factor_);
        boxes[index].height = round(boxes[index].height * scale_factor_);
        res.rectBox = boxes[index];
        res.score = class_scores[index];
        cv::Mat mask_roi = m1(cv::Range(my1, my2), cv::Range(mx1, mx2));
        cv::Mat rm, det_mask;
        cv::resize(mask_roi, rm, cv::Size(boxes[index].width, boxes[index].height),0,0, cv::INTER_CUBIC);
        for (int r = 0; r < rm.rows; r++) {
            for (int c = 0; c < rm.cols; c++) {
                float pv = rm.at<float>(r, c);
                if (pv > 0.5) {
                    rm.at<float>(r, c) = 1.0;
                    int orig_x = boxes[index].x + c;
                    int orig_y = boxes[index].y + r;
                    if (orig_x < 0 || orig_x > src_width_ || orig_y < 0 || orig_y > src_height_)
                    {
                        break;
                    }
                    res.maskPoints.push_back(cv::Point2i(orig_x, orig_y));
                }
                else {
                    rm.at<float>(r, c) = 0.0;
                }
            }
        }
        predectResult_.push_back(res);
    }`

**Comment by glenn-jocher** [comment]
The official script uses `retina_masks=True` for smoother edges. Ensure your deployment code applies similar post-processing techniques.

**Comment by Y-T-G** [comment]
Use the mask for drawing, instead of segments. And don't apply a confidence threshold like in the example.

https://github.com/ultralytics/ultralytics/blob/8154a27e9b332dd2dac7a107003462f0e20f6179/examples/YOLOv8-Segmentation-ONNXRuntime-Python/main.py#L227
